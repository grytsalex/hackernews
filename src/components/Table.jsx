import React, { Component } from 'react';

const isSearched = searchTerm => item =>
    item.title.toLowerCase().includes(searchTerm.toLowerCase());

class Table extends Component {
    render() {
        const { list, pattern, onDismiss } = this.props;

        return(
            <div>
            {list.filter(isSearched(pattern)).map(item =>
            <div key={item.objectID}>
                <span>
                    <a href={item.url}>{item.title}</a>
                </span>
                <span>{item.author}</span>
                <span>{item.num_comments}</span>
                <span>{item.points}</span>
                <span>
                    <button
                        onClick={() => onDismiss(item.objectID)}
                        type="button"
                    >
                    Отбросить
                    </button>
                </span>
            </div>
            )}
            </div>
        )
    }
}

export default Table