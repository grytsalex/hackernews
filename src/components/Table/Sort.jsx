import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSortUp, faSortDown } from "@fortawesome/free-solid-svg-icons";

import { Button } from "../Button";

export const Sort = ({ onSort, sortKey, children, activeSortKey, isSortReverse }) => {
  const sortClass = classNames("button-inline", {
    "button-active": sortKey === activeSortKey,
  });

  return (
        <Button className={sortClass} onClick={() => onSort(sortKey)} style={{}}> 
          {children}
          {isSortReverse && sortKey === activeSortKey ? <FontAwesomeIcon icon={faSortUp} size="lg" /> : <FontAwesomeIcon icon={faSortDown} size="lg" />}
        </Button>
  );
};

Sort.propTypes = {
  children: PropTypes.node,
  sortKey: PropTypes.string,
  onSort: PropTypes.func,
  activeSortKey: PropTypes.string,
  isSortReverse: PropTypes.bool,
};
