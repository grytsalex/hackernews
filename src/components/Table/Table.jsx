import React from "react";
import PropTypes from "prop-types";

import { Sort } from "./Sort";
import { Button } from "../Button";
import { SORT } from "../../helpers";

export const Table = ({ list, onDismiss, sortKey, onSort, isSortReverse }) => {
  const sortedList = SORT[sortKey](list);

  const reverseSortedList = isSortReverse ? sortedList.reverse() : sortedList;

  return (
    <div className="table">
      <div className="table-header">
        <span style={{ width: "40%" }}>
          <Sort sortKey={"TITLE"} onSort={onSort} activeSortKey={sortKey}  isSortReverse={isSortReverse}>
            Title
          </Sort>
        </span>
        <span style={{ width: "30%" }}>
          <Sort sortKey={"AUTHOR"} onSort={onSort} activeSortKey={sortKey} isSortReverse={isSortReverse}>
            Author
          </Sort>
        </span>
        <span style={{ width: "10%" }}>
          <Sort sortKey={"COMMENTS"} onSort={onSort} activeSortKey={sortKey} isSortReverse={isSortReverse}>
            Comments
          </Sort>
        </span>
        <span style={{ width: "10%" }}>
          <Sort sortKey={"POINTS"} onSort={onSort} activeSortKey={sortKey} isSortReverse={isSortReverse}>
            Points
          </Sort>
        </span>
        <span style={{ width: "10%", display: 'table', margin: "auto" }}>Archive</span>
      </div>
      {reverseSortedList.map((item) => (
        <div key={item.objectID} className="table-row">
          <span style={{ width: "40%" }}>
            <a href={item.url}>{item.title}</a>
          </span>
          <span style={{ width: "30%" }}>{item.author}</span>
          <span style={{ width: "10%" }}>{item.num_comments}</span>
          <span style={{ width: "10%" }}>{item.points}</span>
          <span style={{ width: "10%" }}>
            <Button
              onClick={() => onDismiss(item.objectID)}
              className="button-inline"
            >
              Delete
            </Button>
          </span>
        </div>
      ))}
    </div>
  );
};

Table.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      objectID: PropTypes.string.isRequired,
      author: PropTypes.string,
      url: PropTypes.string,
      num_comments: PropTypes.number,
      points: PropTypes.number,
    })
  ).isRequired,
  onDismiss: PropTypes.func.isRequired,
  sortKey: PropTypes.string,
  onSort: PropTypes.func,
};
