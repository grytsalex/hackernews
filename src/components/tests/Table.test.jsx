import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { Table } from '../Table';


Enzyme.configure({ adapter: new Adapter() })

describe('Table', () => {

    const props = {
        list: [
            {title: '1', author: '1', num_comments: 1, points: 1, objectID: 'y'},
            {title: '2', author: '2', num_comments: 1, points: 1, objectID: 'z'}
        ]
    }
    test('should have snapshot', () => {
        const component = renderer.create(<Table {...props}/>);

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })

    it('must shows two items in list', () => {
        const element = shallow(<Table {...props}/>)
        expect(element.find('.table-row').length).toEqual(2)
    })
})