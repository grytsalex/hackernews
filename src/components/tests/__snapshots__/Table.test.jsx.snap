// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should have snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </span>
  </div>
</div>
`;
